lib/services/access_request.go

```diff
@@ -2474,7 +2530,22 @@ func (m *RequestValidator) roleAllowsResource(
 		matchers = append(matchers, NewLoginMatcher(loginHint))
 	}
 	matchers = append(matchers, extraMatchers...)
-	err := roleSet.checkAccess(resource, m.userState.GetTraits(), AccessState{MFAVerified: true}, matchers...)
+	var err error
+	if resource.GetKind() == types.KindSAMLIdPServiceProvider {
+		if types.IsLegacySAMLRBAC(role.GetVersion()) {
+			// legacy roles grant implicit access to saml resource
+			// and can be removed from the reqeust.
+			return false, nil
+		}
```

Comment:
I'm trying to understand how this works, is it true that users with NEVER want to request a v7 role to access a SAMLIdPServiceProvider? I guess the behaviour is that if the user has a single v8+ role then a v7 role will never grant access, and if the user only has v7 roles then they already should have access (unless explicitly denied).